fxa-rust-client/src/lib.rs

```diff
@@ -145,36 +174,198 @@ impl FirefoxAccount {
     pub fn advance(&mut self) {
         let client = FxAClient::new(&self.state.config);
         let state_machine = FxALoginStateMachine::new(client);
-        let state = mem::replace(&mut self.state.login_state, Separated);
+        let state = mem::replace(&mut self.state.login_state, Unknown);
         self.state.login_state = state_machine.advance(state);
     }
 
-    pub fn get_oauth_token(&mut self, scopes: Vec<&str>) -> Result<String> {
-        let scopes_key = self.scopes_key(&scopes);
-        if let Some(cached_token) = self.state.oauth_tokens_cache.get(&scopes_key) {
-            return Ok(cached_token.clone());
+    fn oauth_cache_store(&mut self, info: &OAuthInfo) {
+        let info = info.clone();
+        let scope_key = info.scopes.join("|");
+        self.state.oauth_cache.insert(scope_key, info);
+    }
+
+    fn scope_key_contains_scopes(scope_key: &str, scopes: &[&str]) -> bool {
+        for scope in scopes {
+            if !scope_key.contains(scope) {
```

Comment:
Hmm..this is a simple string search, right?  There could easily be false positives here, e.g. the scope "profile:email" contains that string "profile" but it doesn't actually have that scope.

Reply:
I've corrected that, but in a next PR I'll just translate https://github.com/mozilla/fxa-shared/pull/21/ to Rust and use it.